fem/bilininteg_hdiv.cpp

```diff
@@ -0,0 +1,1987 @@
+   {
+      PADivDivSetup3D(quad1D, ne, ir->GetWeights(), geom->J, coeff, pa_data);
+   }
+   else if (el->GetDerivType() == mfem::FiniteElement::DIV && dim == 2)
+   {
+      PADivDivSetup2D(quad1D, ne, ir->GetWeights(), geom->J, coeff, pa_data);
+   }
+   else
+   {
+      MFEM_ABORT("Unknown kernel.");
+   }
+}
+
+void DivDivIntegrator::AddMultPA(const Vector &x, Vector &y) const
+{
+   if (dim == 3)
+      PADivDivApply3D(dofs1D, quad1D, ne, mapsO->B, mapsC->G,
+                      mapsO->Bt, mapsC->Gt, pa_data, x, y);
+   else if (dim == 2)
+      PADivDivApply2D(dofs1D, quad1D, ne, mapsO->B, mapsC->G,
+                      mapsO->Bt, mapsC->Gt, pa_data, x, y);
+   else
+   {
+      MFEM_ABORT("Unsupported dimension!");
+   }
+}
+
+static void PADivDivAssembleDiagonal2D(const int D1D,
+                                       const int Q1D,
+                                       const int NE,
+                                       const Array<double> &_Bo,
+                                       const Array<double> &_Gc,
+                                       const Vector &_op,
+                                       Vector &_diag)
+{
+   constexpr static int VDIM = 2;
+
+   auto Bo = Reshape(_Bo.Read(), Q1D, D1D-1);
+   auto Gc = Reshape(_Gc.Read(), Q1D, D1D);
+   auto op = Reshape(_op.Read(), Q1D, Q1D, NE);
+   auto diag = Reshape(_diag.ReadWrite(), 2*(D1D-1)*D1D, NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      int osc = 0;
+
+      for (int c = 0; c < VDIM; ++c)  // loop over x, y components
+      {
+         const int D1Dx = (c == 1) ? D1D - 1 : D1D;
+         const int D1Dy = (c == 0) ? D1D - 1 : D1D;
+
+         double div[MAX_Q1D];
+
+         for (int dy = 0; dy < D1Dy; ++dy)
+         {
+            for (int qx = 0; qx < Q1D; ++qx)
+            {
+               div[qx] = 0.0;
+               for (int qy = 0; qy < Q1D; ++qy)
+               {
+                  const double wy = (c == 0) ? Bo(qy,dy) : Gc(qy,dy);
+                  div[qx] += wy * wy * op(qx,qy,e);
+               }
+            }
+
+            for (int dx = 0; dx < D1Dx; ++dx)
+            {
+               for (int qx = 0; qx < Q1D; ++qx)
+               {
+                  const double wx = (c == 0) ? Gc(qx,dx) : Bo(qx,dx);
+                  diag(dx + (dy * D1Dx) + osc, e) += div[qx] * wx * wx;
+               }
+            }
+         }
+
+         osc += D1Dx * D1Dy;
+      }  // loop c
+   });
+}
+
+static void PADivDivAssembleDiagonal3D(const int D1D,
+                                       const int Q1D,
+                                       const int NE,
+                                       const Array<double> &_Bo,
+                                       const Array<double> &_Gc,
+                                       const Vector &_op,
+                                       Vector &_diag)
+{
+   MFEM_VERIFY(D1D <= HDIV_MAX_D1D, "Error: D1D > HDIV_MAX_D1D");
+   MFEM_VERIFY(Q1D <= HDIV_MAX_Q1D, "Error: Q1D > HDIV_MAX_Q1D");
```

Comment:
Good question. In the H(curl) case, we have this check only for 3D Apply functions, not Assemble functions or any 2D functions. My understanding is that this check prevents using too much memory on the GPU, so the 3D case is more restrictive. What I don't understand is that the array sizes depend on the number of elements as well as Q1D and D1D, so is it only the number of DOF's or quadrature points per element that matters, so that one element at a time can be handled? This is a question for @camierjs or @YohannDudouit. Also, should there be a 2D max Q1D and D1D? Should the checks be done also in setup functions or just apply functions?

Reply:
In this particular kernel, only `HDIV_MAX_Q1D` seems to be used. It is used just for the size of a 1D array, so this is not that important here from the point of view of memory usage. The check just makes sure that we do not try to use array entities beyond what is allocated, see the array `double a[HDIV_MAX_Q1D];` below.

Looking at the code below, it caught my attention that when being set `a` is accessed with `a[qx]` with `0<=qx<Q1D` but then it is read using `a[dx]` with `0<=dx<D1Dx`. This seems a little suspicious.

Reply:
@v-dobrev Thanks for finding this. It seems to be a bug, i.e. `a[dx]` should be `a[qx]`. The diagonal preconditioner still works (doesn't have to be perfect), so I didn't catch this. I should add unit tests for the diagonal assembly.